portalnet/src/socket.rs

```diff
@@ -35,3 +45,82 @@ pub fn stun_for_external(local_socket_addr: &SocketAddr) -> Option<SocketAddr> {
         }
     }
 }
+
+pub fn is_local_addr(external_ip: Option<IpAddr>) -> bool {
+    if let Some(external_ip) = external_ip {
+        let network_interfaces = local_ip_address::list_afinet_netifas();
+
+        if let Ok(network_interfaces) = network_interfaces {
+            for (_, ip) in network_interfaces.iter() {
+                if *ip == external_ip {
+                    return true;
+                }
+            }
+        }
+    }
+    false
+}
+
+pub fn upnp_for_external(listen_addr: SocketAddr) -> Option<SocketAddr> {
+    info!("Connecting to UPnP gateway to map local address to external address");
+    let gateway = match igd_next::search_gateway(Default::default()) {
+        Ok(gateway) => gateway,
+        Err(ref err) => {
+            warn!(error = %err, "Error finding gateway");
+            return None;
+        }
+    };
+
+    let local_ip = match TcpStream::connect(gateway.addr) {
```

Comment:
Actually, the previous version of the PR uses local_ip and local_ip_v6 to find the local IP address as below:

```
local_addr = match local_ip_address::local_ip().or(local_ip_address::local_ipv6())local_ip_address::local_ip().or(local_ip_address::local_ipv6()) ...
```

However, as @carver points out that this may be problematic if multiple interfaces exist.  That is why I switch to a TCP way to detect the local IP.

Note that the UPnP protocol also uses an HTTP/TCP connection to the gateway.  Unfortunately, the UPnP library does not return which local IP address is used to talk to the gateway.  So, without modifying UPnP library, making a TCP connection to the gateway is an easy solution to determine the local IP address